### ISSUE TITLE
[rllib] Negative vf_explained_var

### ISSUE THREAD

**Issue by debackerl** [ray-project/ray#10159]
### [rllib] Negative vf_explained_var

Python 3.8.5, TensorFlow 2.2.0, rllib 0.8.6, Linux Pop_OS 20.04

I have been trying PPO on a simple experiment. If first reward after an action is too negative, then it looks like that it won't realize that future positive rewards will make up for it. What I observed is that the vf_explained_var will first briefly grow to 0.8 before reversing, and oscillating back and forth between 0 and -1 (and always stay negative indeed). In the end, it learns to not do anything, so that there is no large negative reward, but also no positive ones. Would that be a limitation in PPO or am I formulating an ill problem?

In a nutshell, the environment has organisms which are placed in a cold storage and a live pool. Population in cold storage is stable, but goes up and down in the live pool depending on the current climate. The action specifies how much to transfer between those two, with a loss of 10% of organism during transfer in either direction. The growth rate of the live pool depending on time is given by a simple formula: exp(2*(0.02*sin(t)+0.1*sin(t*0.1)+sin(t*0.005))).

The episode lasts at most 252 timesteps. A good policy would transfer all organisms in the live pool when the environment makes them replicate, and transfer them back to cold storage before the environment kills most of them. sin(t*0.005) will be the main growth driver.

Does a negative vf_explained_var means that the expected value function is negatively correlated to the actual reward?

Thank you for any hint!!

Code:
```
import math
import gym
import numpy as np
from gym import spaces
from random import Random

import ray
from ray import tune
from ray.tune.registry import register_env
from ray.rllib.agents.ppo import DEFAULT_CONFIG as PPO_DEFAULT_CONFIG

class TraderEnv(gym.Env):
  def __init__(self, env_config):
    self.action_space = spaces.Box(low=-1.0, high=1.0, shape=(1,), dtype=np.float64)
    self.observation_space = spaces.Box(low=np.array([-1.0, 0.0, -1.0, -10.0, -1, -1, -1]), high=np.array([2.0, 1.0, 1.0, 15, 1, 1, 1]), shape=(7,), dtype=np.float64)

    self._transfer_loss = env_config.get('transfer_loss', 0.0) # cost = transfer_loss * transfer_amount
    self._max_days = env_config.get('max_days', 252)
    self._shutdown_population = env_config.get('shutdown_population', 0.1)

    self.seed(0)

  def seed(self, seed=0):
    self._rnd = Random(seed)
    self._phase = self._rnd.random() * 800
    return [ seed ]

  def reset(self):
    self._t = t = self._phase
    self._log_population_t = self._log_population(t)
    self._cold_storage = 1.0
    self._live = 0.0
    self._perf = 0.0

    return [ self._perf, self._live / (self._cold_storage + self._live), self._log_population_t - self._log_population(t-1), self._log_population_t, math.sin(t), math.sin(t*0.1), math.sin(t*0.005) ]

  def step(self, action):
    self._t += 1
    t = self._t

    prev_log_population_t = self._log_population_t
    self._log_population_t = log_population_t = self._log_population(t)

    operation = action[0]

    prev_total = self._cold_storage + self._live

    if operation > 0.0:
      # transfer from cold storage to live population
      amount = self._cold_storage * operation
      self._cold_storage -= amount
      # amount = live + transfer_loss * live <=> amount = (1 + transfer_loss) * live <=> live = amount / (1 + transfer_loss)
      self._live += amount / (1 + self._transfer_loss)
    elif operation < 0.0:
      # transfer from live population to cold storage
      amount = self._live * -operation
      self._live -= amount
      self._cold_storage += amount * (1 - self._transfer_loss)

    cost = self._cold_storage + self._live - prev_total

    growth = math.exp(log_population_t - prev_log_population_t) # = math.exp(log_population_t) / math.exp(prev_log_population_t) = population_t / prev_population_t
    self._live *= growth

    total = self._cold_storage + self._live

    reward = math.log(total / prev_total) # = math.log(self._total) - math.log(prev_total)
    self._perf = (self._perf + math.log((total + cost) / prev_total)) * 0.5 # how good the agent has been doing recently

    observation = [ self._perf, self._live / (self._cold_storage + self._live), log_population_t - prev_log_population_t, log_population_t, math.sin(t), math.sin(t*0.1), math.sin(t*0.005) ]

    done = total < self._shutdown_population or (self._t - self._phase) >= self._max_days

    info = {}

    return observation, reward, done, info

  def _log_population(self, t):
    return 2*(0.02*math.sin(t)+0.1*math.sin(t*0.1)+math.sin(t*0.005))

if __name__ == "__main__":
  ray.init(num_cpus=8, memory=16*1024*1024*1024, object_store_memory=4*1024*1024*1024)

  trainer_config = PPO_DEFAULT_CONFIG.copy()

  trainer_config['env'] = 'Reactor'
  trainer_config['env_config'] = {
    'transfer_loss': 0.1, # 10% loss during transfer in both directions
  }

  trainer_config['num_gpus'] = 0 # can use fractions to have multiple workers in parallel
  trainer_config['num_workers'] = 0
  trainer_config['num_envs_per_worker'] = 16

  trainer_config['framework'] = 'tf'
  trainer_config["sgd_minibatch_size"] = 50
  trainer_config["train_batch_size"] = 400
  trainer_config["num_sgd_iter"] = 10

  register_env("Reactor", lambda env_config: TraderEnv(env_config))

  stop = {
    #"time_total_s": 600.0,
    #"training_iteration": 20,
    "timesteps_total": 500000,
    #"episode_reward_mean": 0.116
  }

  analysis = tune.run('PPO', name="Reactor", num_samples=16, config=trainer_config, checkpoint_at_end=True, stop=stop)

  print("Best config: ", analysis.get_best_config(metric='episode_reward_mean'))

  ray.shutdown()
```
